test/infrastructure/container/docker.go

```diff
@@ -383,11 +383,12 @@ func (d *docker) RunContainer(ctx context.Context, runConfig *RunContainerInput,
 		// privileges, but this flag also changes some mounts that are necessary
 		// including some ones docker would otherwise do by default.
 		// for now this is what we want. in the future we may revisit this.
-		Privileged:   true,
-		SecurityOpt:  []string{"seccomp=unconfined"}, // ignore seccomp
-		NetworkMode:  dockercontainer.NetworkMode(runConfig.Network),
-		Tmpfs:        runConfig.Tmpfs,
-		PortBindings: nat.PortMap{},
+		Privileged:    true,
+		SecurityOpt:   []string{"seccomp=unconfined"}, // ignore seccomp
+		NetworkMode:   dockercontainer.NetworkMode(runConfig.Network),
+		Tmpfs:         runConfig.Tmpfs,
+		PortBindings:  nat.PortMap{},
+		RestartPolicy: dockercontainer.RestartPolicy{Name: "unless-stopped"},
```

Comment:
Q: what is the CAPD behaviour when a container is stopped? Is it going to fail (and possibly surface and error) or is it going to recreate a new container?

Reply:
This may depend on when and how the stop is done.

If the entire service is stopped, then everything goes down and is happy when it starts back up.

Testing locally once everything is in a steady state, it looks like things are still happy with just stopping the MachineDeployment containers while leaving others running. I think this may be because we don't actually look at the Status of the container if it's been found, but we are setting the actual status here:

https://github.com/kubernetes-sigs/cluster-api/blob/c90f3763617d8bd33c697bebd1be5cfe37a2102a/test/infrastructure/docker/docker/util.go#L98-L103

Stopping the controlplane container does result in these errors in the logs:

```
capi-kubeadm… │ E0727 19:51:03.476496       8 reflector.go:138] pkg/mod/k8s.io/client-go@v0.21.2/tools/cache/reflector.go:167: Failed to watch *v1.ClusterRole: failed to list *v1.ClusterRole: context deadline exceeded
```

But again we just keep going.

I would imagine if a container is stopped at some point during initial deployment that it could cause an issue.

It's a really good question that we should probably think about. We may want to try to reconcile state if we find one of the containers has an "exited" status. Let me know if you think I should file an issue for that. I think it's a separate concern as it's a behavior we have today even without this change.

So at least with this, if someone has not manually stopped a container, a restart will get it all back. The other two restartPolicy options are `always` and `on-failure`. `on-failure` would only restart if the container exited with a non-0 exit code, so that would not change the behavior if someone were to manually stop the container. And `always` only would restart the container on Docker Engine restart if it was manually stopped.